ApiData.ts

```diff
@@ -0,0 +1,264 @@
+	lastVal: number = -1;
+	maxRange: number;
+
+	constructor(pageSize: number) {
+		this.pageSize = pageSize;
+		this.maxRange = 0;
+	}
+
+	/** Initialize method to set up the grid */
+	initialize(): Promise<void> {
+		this.adjustGridHeight();
+		return this.recordCount()
+			.then(() => this.fetchColumns())
+			.then(() => this.fetchAndDisplayRecords())
+			.then(() => this.setupControls());
+	}
+
+	/** Fetch total record count from the server,fetches data from an API and populates class properties */
+	recordCount(): Promise<void> {
+		return this.fetchData('http://localhost:2050/recordCount')
+			.then((response: number | string) => {
+				const totalItems = <number>response;
+				this.totalItems = totalItems;
+			})
+			.catch(error => {
+				console.error('Failed to fetch record count:', error);
+				alert('Failed to fetch record count.');
+				throw error;
+			});
+	}
+
+	/** Use the fetchData() func to make an HTTP request to the API endpoint and process the data */
+	fetchColumns(): Promise<void> {
+		return this.fetchData('http://localhost:2050/columns')
+			.then((response: number | string) => {
+				const res = JSON.parse(<string>(response));
+				this.columnNames = res.map((columnName: string) => ({ name: columnName }));
+				// Initialize the 'data' property as an empty array of GridData objects
+				this.data = new Array<GridData>(this.columnNames.length);
+			})
+			.catch(() => {
+				throw ('Failed to fetch columns.');
+			});
+	}
+
+	/** Get records from API for fetch and search functionality */
+	fetchAndProcessRecords(from: number, to: number): Promise<GridData[]> {
+		$('#spinner').show();
+		$('#grid').hide();
+
+		return this.fetchData(`http://localhost:2050/records?from=${from}&to=${to}`)
+			.then((response: number | string) => {
+				const res = JSON.parse(<string>(response));
+				const processedData = res.map((record: string) => {
+					const obj: GridData = {};
+					for (let j = 0; j < this.columnNames.length && j < record.length; j++) {
+						obj[this.columnNames[j].name] = record[j];
+					}
+					return obj;
+				});
+				$('#spinner').hide();
+				$('#grid').show();
+				return processedData;
+			})
+			.catch(() => {
+				throw ('Failed to fetch records');
+			});
+	}
+
+	/** Fetches records using fetchAndProcessRecords(), processes them, displays them, and updates page information. */
+	fetchAndDisplayRecords(): Promise<void> {
+		this.maxRange = this.totalItems - 1;
+		let from = this.firstVal;
+		let to = Math.min(from + this.pageSize, this.maxRange);
+
+		if (to >= this.maxRange) {
+			const lastPage = Math.floor(this.maxRange / this.pageSize) + 1;
+			this.currentPage = lastPage;
+			from = this.maxRange - this.pageSize;
+			to = this.maxRange;
+		}
+
+		return this.fetchAndProcessRecords(from, to)
+			.then(processedData => {
+				this.data = processedData;
+				this.displayRecords();
+				this.updatePageInfo();
+			})
+			.catch(error => {
+				console.error('Failed to fetch records:', error);
```

Comment:
Technically speaking, what do you think your `console.error` will look like since you have nested promises?
Think about it 🤔